src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/WorkspaceContextHost.WorkspaceContextHostInstance.cs

```diff
@@ -0,0 +1,115 @@
+using Microsoft.VisualStudio.ProjectSystem.Utilities;
+using Microsoft.VisualStudio.Threading;
+
+namespace Microsoft.VisualStudio.ProjectSystem.LanguageServices
+{
+    internal partial class WorkspaceContextHost
+    {
+        /// <summary>
+        ///     Responsible for creating and initializing <see cref="IWorkspaceProjectContext"/> and sending 
+        ///     on changes to the project to the <see cref="IApplyChangesToWorkspaceContext"/> service.
+        /// </summary>
+        internal partial class WorkspaceContextHostInstance : AbstractMultiLifetimeInstance
+        {
+            private readonly ConfiguredProject _project;
+            private readonly IProjectSubscriptionService _projectSubscriptionService;
+            private readonly IProjectThreadingService _threadingService;
+            private readonly IUnconfiguredProjectTasksService _tasksService;
+            private readonly Lazy<IWorkspaceProjectContextProvider> _workspaceProjectContextProvider;
+            private readonly ExportFactory<IApplyChangesToWorkspaceContext> _applyChangesToWorkspaceContextFactory;
+
+            private IWorkspaceProjectContext _context;
+            private ExportLifetimeContext<IApplyChangesToWorkspaceContext> _applyChangesToWorkspaceContext;
+            private readonly DisposableBag _subscriptions = new DisposableBag(CancellationToken.None);
+
+            public WorkspaceContextHostInstance(ConfiguredProject project,
+                                                IProjectThreadingService threadingService,
+                                                IUnconfiguredProjectTasksService tasksService,
+                                                IProjectSubscriptionService projectSubscriptionService,
+                                                Lazy<IWorkspaceProjectContextProvider> workspaceProjectContextProvider,
+                                                ExportFactory<IApplyChangesToWorkspaceContext> applyChangesToWorkspaceContextFactory)
+                : base(threadingService.JoinableTaskContext)
+            {
+                _project = project;
+                _projectSubscriptionService = projectSubscriptionService;
+                _threadingService = threadingService;
+                _tasksService = tasksService;
+                _workspaceProjectContextProvider = workspaceProjectContextProvider;
+                _applyChangesToWorkspaceContextFactory = applyChangesToWorkspaceContextFactory;
+            }
+
+            protected override async Task InitializeCoreAsync(CancellationToken cancellationToken)
+            {
+                IWorkspaceProjectContext context = await _workspaceProjectContextProvider.Value.CreateProjectContextAsync(_project)
+                                                                                               .ConfigureAwait(true);
+
+                if (context == null)
+                    return;
+
+                await TaskScheduler.Default;
```

Comment:
Please add a comment that calls out that we want this work to happen on a background thread and why that is.